File: mobile/lib/utils/github_project_identity.dart
import '../models/project_item.dart';

String normalizeGithubRepoFullName(String fullName) {
  return fullName
      .trim()
      .split('/')
      .map((s) => s.trim())
      .where((s) => s.isNotEmpty)
      .join('/');
}

String githubRepoPublicationUrl(String fullName) {
  final normalized = normalizeGithubRepoFullName(fullName);
  final parts = normalized.split('/');
  final path = parts.map((p) => p.toLowerCase()).join('/');
  return 'https://github.com/$path';
}

bool projectMatchesGithubRepo(ProjectItem project, String fullName) {
  if (!project.isGithub) return false;
  final want = normalizeGithubRepoFullName(fullName).toLowerCase();
  if (!want.contains('/')) return false;

  final repoFullName = project.repoFullName?.trim();
  if (repoFullName != null && repoFullName.isNotEmpty) {
    if (normalizeGithubRepoFullName(repoFullName).toLowerCase() == want) {
      return true;
    }
  }

  final url = project.publicationUrl?.trim().toLowerCase().replaceAll(RegExp(r'/+$'), '');
  if (url != null && url.isNotEmpty) {
    return url == githubRepoPublicationUrl(fullName).toLowerCase();
  }
  return false;
}

bool projectListContainsGithubRepo(List<ProjectItem> projects, String fullName) {
  for (final p in projects) {
    if (projectMatchesGithubRepo(p, fullName)) return true;
  }
  return false;
}

List<ProjectItem> removeGithubRepoFromProjects(
  List<ProjectItem> projects,
  String repoFullName,
) {
  return projects.where((p) => !projectMatchesGithubRepo(p, repoFullName)).toList();
}

File: webapp/src/lib/profile/githubProjectIdentity.ts
export function normalizeGithubRepoFullName(fullName: string): string {
  return fullName
    .trim()
    .split("/")
    .map((s) => s.trim())
    .filter(Boolean)
    .join("/");
}
export function githubRepoPublicationUrl(fullName: string): string {
  const n = normalizeGithubRepoFullName(fullName);
  const parts = n.split("/");
  const path = parts.map((s) => s.toLowerCase()).join("/");
  return `https://github.com/${path}`;
}
export function projectMatchesGithubRepo(
  project: unknown,
  fullName: string,
): boolean {
  const want = normalizeGithubRepoFullName(fullName).toLowerCase();
  if (!want.includes("/")) return false;
  const p = project as {
    source?: string;
    repoFullName?: string;
    publicationUrl?: string;
  };
  if (p.source !== "github") return false;
  if (
    typeof p.repoFullName === "string" &&
    normalizeGithubRepoFullName(p.repoFullName).toLowerCase() === want
  ) {
    return true;
  }
  if (typeof p.publicationUrl === "string") {
    const u = p.publicationUrl.trim().toLowerCase().replace(/\/+$/, "");
    return u === githubRepoPublicationUrl(fullName).toLowerCase();
  }
  return false;
}
